File: MiniGame-NET/MiniGameEngine/Text.cs
// Copyright 2023 Peter Bakota
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Runtime.InteropServices;
using MiniGameEngine.Utils;
using SDL2;

namespace MiniGameEngine
{
    public static class Text
    {
        public static readonly int FONT_SIZE = 16;
        public static readonly int FONT_COUNT = 74;
        private const int COLOR_INDEX = 31;
        private const int SHADOW_INDEX = 25;
        private const int TRANSPARENT_INDEX = 0;
        private static readonly byte[] _fontData;

        static Text()
        {
            var font = SDLUtil.CheckSDLErr(() => SDL.SDL_LoadBMP(Path.Combine(AssetsLoader._cwd, AssetsLoader._assetsFolder, "Fonts", "font.bmp")));

            if (SDL.SDL_MUSTLOCK(font))
                SDLUtil.CheckSDLErr(() => SDL.SDL_LockSurface(font));

            var surface = Marshal.PtrToStructure<SDL.SDL_Surface>(font);

            _fontData = new byte[FONT_SIZE * FONT_SIZE * FONT_COUNT];
            var dptr = 0;

            for (var c = 0; c < FONT_COUNT; ++c)
            {
                var ptr = surface.pixels + FONT_SIZE * c;
                for (var y = 0; y < FONT_SIZE; ++y)
                {
                    for (var x = 0; x < FONT_SIZE; ++x)
                    {
                        _fontData[dptr++] = Marshal.ReadByte(ptr);
                        ptr += sizeof(byte);
                    }
                    ptr += surface.pitch - FONT_SIZE * sizeof(byte);
                }
            }

            if (SDL.SDL_MUSTLOCK(font))
                SDL.SDL_UnlockSurface(font);

            SDL.SDL_FreeSurface(font);
        }

        private static void DrawToCharBuffer(IntPtr buffer, int pitch, uint color, uint shadow, int chr)
        {
            var offset = FONT_SIZE * FONT_SIZE * (chr - 0x21);
            var ptr = buffer;

            for (var y = 0; y < FONT_SIZE; ++y)
            {
                for (var x = 0; x < FONT_SIZE; ++x)
                {
                    switch (_fontData[offset++])
                    {
                        case COLOR_INDEX:
                            Marshal.WriteIntPtr(ptr, (IntPtr)color);
                            break;
                        case SHADOW_INDEX:
                            Marshal.WriteIntPtr(ptr, (IntPtr)shadow);
                            break;
                        case TRANSPARENT_INDEX:
                            // Do nothing here!
                            break;
                    }
                    ptr += sizeof(uint);
                }
                ptr += pitch - FONT_SIZE * sizeof(uint);
            }
        }

        private static void DrawChar(SDL.SDL_Surface surface, int x, uint color, uint shadow, int chr)
            => DrawToCharBuffer(surface.pixels + x * FONT_SIZE * sizeof(uint), surface.pitch, color, shadow, chr);

        private static void DrawText(IntPtr surface, string text, uint color, uint shadow)
        {
            if (SDL.SDL_MUSTLOCK(surface))
                SDLUtil.CheckSDLErr(() => SDL.SDL_LockSurface(surface));

            var surface2 = Marshal.PtrToStructure<SDL.SDL_Surface>(surface);
            var x = 0;
            foreach (var c in text.ToUpper())
                if (c == 0x20) x++; else DrawChar(surface2, x++, color, shadow, c);

            if (SDL.SDL_MUSTLOCK(surface))
                SDL.SDL_UnlockSurface(surface);
        }

        public static void DrawText(IntPtr renderer,
            int x,
            int y,
            SDL.SDL_Color color,
            SDL.SDL_Color shadow,
            string text, double size = 1.0f, double angle = 0.0f, bool vertFlip = false, bool horizFlip = false)
        {
            var surface = SDL.SDL_CreateRGBSurface(0, text.Length * FONT_SIZE, FONT_SIZE, 32, 0xff000000, 0x00ff0000, 0x0000ff00, 0x000000ff);
            DrawText(surface, text, color.ToRGBA(), shadow.ToRGBA());

            var texture = SDL.SDL_CreateTextureFromSurface(renderer, surface);
            var dst = SDLUtil.Rect(x, y, (int)(text.Length * FONT_SIZE * size), (int)(FONT_SIZE * size));

            var flip = (horizFlip ? SDL.SDL_RendererFlip.SDL_FLIP_HORIZONTAL : SDL.SDL_RendererFlip.SDL_FLIP_NONE) |
                       (vertFlip ? SDL.SDL_RendererFlip.SDL_FLIP_VERTICAL : SDL.SDL_RendererFlip.SDL_FLIP_NONE);

            SDLUtil.CheckSDLErr(() => SDL.SDL_RenderCopyEx(renderer, texture, IntPtr.Zero, ref dst, angle, IntPtr.Zero, flip));

            SDL.SDL_DestroyTexture(texture);
            SDL.SDL_FreeSurface(surface);
        }
    }

}

File: MiniGame-Go/engine/text.go
// Copyright 2023 Peter Bakota
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package engine

import (
	"strings"

	"github.com/veandco/go-sdl2/sdl"
)

const (
	FONT_SIZE         = 16
	FONT_COUNT        = 74
	COLOR_INDEX       = 31
	SHADOW_INDEX      = 25
	TRANSPARENT_INDEX = 0
)

type Text struct {
	fontData []byte
}

// NOTE: OMG, this approach with Self-referential functions to implement default values is so bad, but this is the "Go way" I presume,
// since it was recommended by the author of Go: https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html
type TextDrawOptions struct {
	Size  float64
	Angle float64
	Vflip bool
	Hflip bool
}

type Option func(f *TextDrawOptions)

func WithSize(size float64) Option {
	return func(f *TextDrawOptions) {
		f.Size = size
	}
}

func WithAngle(angle float64) Option {
	return func(f *TextDrawOptions) {
		f.Angle = angle
	}
}

func WithVflip(flip bool) Option {
	return func(f *TextDrawOptions) {
		f.Vflip = flip
	}
}

func WithHflip(flip bool) Option {
	return func(f *TextDrawOptions) {
		f.Hflip = flip
	}
}

func NewTextDrawOptions(opts ...Option) *TextDrawOptions {
	options := &TextDrawOptions{
		Size:  1,
		Angle: 0,
		Vflip: false,
		Hflip: false,
	}
	for _, applyOpt := range opts {
		applyOpt(options)
	}

	return options
}

func NewText(loader *Loader) *Text {
	p := &Text{}

	font := loader.LoadFont("assets/Fonts/font.bmp")
	defer font.Free()

	if font.MustLock() {
		font.Lock()
	}

	p.fontData = make([]byte, FONT_SIZE*FONT_SIZE*FONT_COUNT)
	pixels := font.Pixels()

	offset := 0
	for c := 0; c < FONT_COUNT; c++ {
		for y := 0; y < FONT_SIZE; y++ {
			for x := 0; x < FONT_SIZE; x++ {
				p.fontData[offset] = pixels[FONT_SIZE*c+y*int(font.Pitch)+x]
				offset++
			}
		}
	}

	if font.MustLock() {
		font.Unlock()
	}
	return p
}

func (t *Text) drawToCharBuffer(pixels []byte, col int, pitch int32, color uint32, shadow uint32, chr byte) {
	offset := FONT_SIZE * FONT_SIZE * int(chr-0x21)
	for y := 0; y < FONT_SIZE; y++ {
		for x := 0; x < FONT_SIZE; x++ {
			switch t.fontData[offset] {
			case COLOR_INDEX:
				// ARGB
				pixels[col+y*int(pitch)+x*4+0] = (byte)(color & 0xff)
				pixels[col+y*int(pitch)+x*4+1] = (byte)(color >> 8 & 0xff)
				pixels[col+y*int(pitch)+x*4+2] = (byte)(color >> 16 & 0xff)
				pixels[col+y*int(pitch)+x*4+3] = (byte)(color >> 24 & 0xff)
			case SHADOW_INDEX:
				// ARGB
				pixels[col+y*int(pitch)+x*4+0] = (byte)(shadow & 0xff)
				pixels[col+y*int(pitch)+x*4+1] = (byte)(shadow >> 8 & 0xff)
				pixels[col+y*int(pitch)+x*4+2] = (byte)(shadow >> 16 & 0xff)
				pixels[col+y*int(pitch)+x*4+3] = (byte)(shadow >> 24 & 0xff)
			case TRANSPARENT_INDEX:
				// Do nothing here
			}
			offset++
		}
	}
}

func (t *Text) drawChar(surface *sdl.Surface, x int, color uint32, shadow uint32, chr byte) {
	t.drawToCharBuffer(surface.Pixels(), x*FONT_SIZE*4, surface.Pitch, color, shadow, chr)
}

func (t *Text) drawText(surface *sdl.Surface, text string, color uint32, shadow uint32) {
	if surface.MustLock() {
		surface.Lock()
	}

	x := 0
	text = strings.ToUpper(text)
	for c := 0; c < len(text); c++ {
		b := text[c]
		if b == 0x20 {
			x++
		} else {
			t.drawChar(surface, x, color, shadow, b)
			x++
		}
	}

	if surface.MustLock() {
		surface.Unlock()
	}
}

func (t *Text) DrawText(renderer *sdl.Renderer, x, y int32, color *sdl.Color, shadow *sdl.Color, text string) {
	t.DrawTextEx(renderer, x, y, color, shadow, text, &TextDrawOptions{
		Size:  1.0,
		Angle: 0.0,
		Vflip: false,
		Hflip: false,
	})
}

func (t *Text) DrawTextEx(renderer *sdl.Renderer, x, y int32, color *sdl.Color, shadow *sdl.Color, text string, options *TextDrawOptions) {
	surface, err := sdl.CreateRGBSurface(0, int32(len(text)*FONT_SIZE), FONT_SIZE, 32, 0xff000000, 0x00ff0000, 0x0000ff00, 0x000000ff)
	if err != nil {
		panic(err)
	}
	defer surface.Free()

	t.drawText(surface, text, color.Uint32(), shadow.Uint32())
	texture, err := renderer.CreateTextureFromSurface(surface)
	if err != nil {
		panic(err)
	}
	defer texture.Destroy()

	dst := &sdl.Rect{
		X: x,
		Y: y,
		W: (int32)(float64(len(text)) * FONT_SIZE * options.Size),
		H: (int32)(FONT_SIZE * options.Size),
	}

	flip := sdl.FLIP_NONE
	if options.Vflip {
		flip |= sdl.FLIP_VERTICAL
	}
	if options.Hflip {
		flip |= sdl.FLIP_HORIZONTAL
	}

	renderer.CopyEx(texture, nil, dst, options.Angle, nil, flip)
}
